### PR TITLE
support nebula source for flink sql connector

### PR DESCRIPTION
**Pull request by liuxiaocs7**
try to add flink sql support, it's still under development.

### CODE REVIEW

**Comment by Nicole00** [comment]
Please see the conflicts, sorry for these conflicts because of another similar work which has been merged.

**Comment by liuxiaocs7** [comment]
related to #58.

UT result are as follow:

source vertex:

```
7> +I[63, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
14> +I[66, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
13> +I[64, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
6> +I[65, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
12> +I[68, aba, 张三, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, true, 1.2, 1.0, 03:12:12, POLYGON((0.0 1.0,1.0 2.0,2.0 3.0,0.0 1.0))]
2> +I[62, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
1> +I[67, 李四, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, true, 1.2, 1.0, 03:12:12, POLYGON((0.0 1.0,1.0 2.0,2.0 3.0,0.0 1.0))]
3> +I[61, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
```
source edges:

```
3> +I[61, 62, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
1> +I[63, 64, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
2> +I[62, 63, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, POINT(1.0 3.0)]
15> +I[65, 66, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
13> +I[67, 68, 0, 李四, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, true, 1.2, 1.0, 03:12:12, POLYGON((0.0 1.0,1.0 2.0,2.0 3.0,0.0 1.0))]
11> +I[68, 61, 0, aba, 张三, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, true, 1.2, 1.0, 03:12:12, POLYGON((0.0 1.0,1.0 2.0,2.0 3.0,0.0 1.0))]
14> +I[66, 67, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
16> +I[64, 65, 0, aba, abcdefgh, 1, 1111, 22222, 6412233, 2019-01-01, 2019-01-01T04:12:12, 435463424, false, 1.2, 1.0, 03:12:12, LINESTRING(1.0 3.0,2.0 4.0)]
```

File: connector/src/main/java/org.apache.flink/connector/nebula/utils/WriteModeEnum.java [hunk 1]
```diff
@@ -26,4 +26,19 @@ public enum WriteModeEnum {
     WriteModeEnum(String mode) {
         this.mode = mode;
     }
+
+    public static boolean checkValidWriteMode(String modeName) {
+        return chooseWriteMode(modeName) != INSERT
+                || INSERT.name().equalsIgnoreCase(modeName);
```
Comment (Nicole00) [inline connector/src/main/java/org.apache.flink/connector/nebula/utils/WriteModeEnum.java:32]: what's the purpose for this function?  always return true?

File: connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (liuxiaocs7) [reply connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java:1]: Thanks for the reminder, indeed, I will revise this.

File: connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    private NebulaDeserializationConverter createInternalConverter(LogicalType type) {
+        switch (type.getTypeRoot()) {
+            case NULL:
+                return val -> null;
+            case BOOLEAN:
+                return ValueWrapper::asBoolean;
+            case TINYINT:
+            case SMALLINT:
+            case INTEGER:
+                return val -> (int) val.asLong();
```
[the remark this answers is not in the document]
Comment (liuxiaocs7) [reply connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java:122]: Done.

File: connector/src/main/java/org.apache.flink/connector/nebula/utils/WriteModeEnum.java (hunk 1 above)
Comment (liuxiaocs7) [reply connector/src/main/java/org.apache.flink/connector/nebula/utils/WriteModeEnum.java:32]: This function is going to be used to verify the `write-mode` parameter passed in through `with` in flink sql. This Function returns true if the mode parameter is one of INSERT, UPDATE, DELETE(ignoreCase), return false if it is any other string, such `'write-mode' = 'add'`.

```java
public static boolean checkValidWriteMode(String modeName) {
    return chooseWriteMode(modeName) != INSERT
            || INSERT.name().equalsIgnoreCase(modeName);
}

public static WriteModeEnum chooseWriteMode(String modeName) {
    if (UPDATE.name().equalsIgnoreCase(modeName)) {
        return UPDATE;
    }
    if (DELETE.name().equalsIgnoreCase(modeName)) {
        return DELETE;
    }
    return INSERT;
}
```

But now the `write-mode` has been removed from with clause, and it seems better to use the `enumType` to validate that comes with flink, so this part of the logic can be removed.

File: connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (liuxiaocs7) [reply connector/src/main/java/org.apache.flink/connector/nebula/table/NebulaRowDataConverter.java]: Now after the modification, all integer types now return Long. The converter works fine. However, if the table declaration is INT, the RowDataSerializer(use to serialize RowData to another compute operator) will generate an exception, that is, the integer of type Long cannot be cast to Integer. The current temporary solution is to directly set all types in the table declaration to BIGINT.

**Review by Nicole00 — APPROVED** [review]

**Pull request merged**